### ISSUE TITLE
Clarify SED upper limits

### ISSUE THREAD

**Issue by cdeil** [open-gamma-ray-astro/gamma-astro-data-formats#75]
@woodmd - It's not clear to me yet how to store upper limits in the SED format.

The spec (see [here](http://gamma-astro-data-formats.readthedocs.io/en/latest/results/flux_points/index.html#error-columns)) doesn't explicitly say what values to fill for upper limits and the example files don't contain upper limits.

In Gammapy we currently don't use an `UL` column, but instead a `NaN` value in the down error column to store an UL (see [here](https://gammapy.readthedocs.io/en/latest/api/gammapy.spectrum.DifferentialFluxPoints.html)).
I think (but didn't double-check, the Fermi-LAT catalogs also don't have an `UL` column, but declare ULs by `NULL` value in the lower error. This is just to mention examples, I'm not saying this has an influence on what we specify here.
@zblz proposed an extra bool `ul` flag column to say which point is which, instead of "magic values" (see [here](http://naima.readthedocs.io/en/latest/dataformat.html)).

At this point I don't have an opinion on which UL point serialisation option is best. Is there an advantage (e.g. simpler storage or processing) to not introducing an `UL` colum, but to re-use the lower / upper error columns?
If we keep the UL column, should we require some special values like `NaN` or `NULL` in the other "flux" columns, or should their content be undefined / irrelevant for ULs?

Are there use cases for "lower limits", i.e. should we treat "upper limits" in a special way or just have a system that works for symmetric confidence intervals?

@adonath or @joleroi or @registerrier @giacomov  - Do you have a preference here?

Maybe this?
```
flux flux_err flux_ul
val val nan # not an UL
nan nan val # is an UL
```
Or this, to still allow storing the maximum likelihood estimate, even for UL?
```
flux flux_err flux_ul
val val nan # not an UL
val val val # is an UL
```

@woodmd - Do you have time to make a pull request for this soon?
Or should I make it?
Please say your preference in any case.

(We need this for `gamma-cat`, so if possible, I'd like to decide and implement this in the spec and update the examples to show UL cases in the next days.)

**Comment by woodmd** [comment]
In fermipy we compute ULs, values, and TS for all bins and then allow the user to choose the between the UL and value according to the TS.  In my view having both UL and value for every bin is useful since there are circumstances where you may want to use the UL even if the bin is significant or use the value for low significance bins.  This also gives extra flexibility to choose a different significance/TS cut for selecting between the value and UL representations.

That said if you prefer to explicitly say whether a column should be treated as an UL then I think i prefer the bool column solution over the magic value one.  I would be happy with a `ul` column as proposed by @zblz.

I doubt that there are many applications for lower limits but for completeness maybe we should go ahead and add it to the spec as well.  The semantics would basically be identical to the UL column.

I can put together a PR with these changes and also update the examples to include ULs.  Do we want the `diff_flux_points` and `int_flux_points` SED types to require an UL column?

**Comment by cdeil** [comment]
> That said if you prefer to explicitly say whether a bin should be treated as an UL then I think i prefer the bool column solution over the magic value one.

It's not really about my preference. There's many papers that publish an UL, and I think the SED format should support storing published SED data. Example SED ECSV I wrote just now: https://github.com/gammapy/gamma-cat/pull/25

Agreed that it's a valid use case that we should support?

> I would be happy with a ul column as proposed by @zblz.

OK.
As I said, I don't have a preference yet (maybe I will after writing many such SED files by hand, probably not).

> I can put together a PR with these changes and also update the examples to include ULs. 

Thanks!!

> Do we want the diff_flux_points and int_flux_points SED types to require an UL column?

As I've said previously, I'm not sure how useful those types are, so I don't really care.
As far as I'm concerned any file that has`e_ref, dnde` is of type "diff flux point", no matter the details on the error or limit info.

**Comment by woodmd** [comment]
I've thought about this a bit more and I think we'll need a convention for null values in addition to the UL column.  The issue is that we need to support the case where both value and ULs are defined (such as in SED files produced by fermipy) as well as cases where only UL or value is defined (such as the archival measurements you use in your example).  I've just opened #76 to add both to the specification.

**Comment by cdeil** [comment]
This discussion has concluded and the solution was implemented by @woodmd in #76.
Closing this issue now.

@woodmd - Thank you!!!
